### PR TITLE
fix Misstep

### LINKED ISSUE #2455: training_epoch_end log output gets combined with next epoch training
## 🐛 Bug

So, I put 'training_epoch_end' function in my LightningModule. I have it return this dictionary
`{'log':{'train_loss': tensor(0.3616, device='cuda:0'), 'epoch': 0}`

I check the `run_training_epoch_end` function in the Pytorch library, it looks like it is working normally as `log_epoch_metrics` is showing the 'log' part in the dictionary produced by 'training_epoch_end' function
`{'train_loss': tensor(0.3616, device='cuda:0'), 'epoch': 0}`

So, they send it off to the logger. But there is problem, it is trying to combine the dictionary above with the results from the training step fo the next epoch. When I check the variable `self._metrics_to_agg`, I get the following result. Of course, it is impossible to combine these dictionaries as they have different keys. I guess the main problem is that the code is combining the log results of `run_training_epoch_end` function with the results of the next training batch.
`[{'train_loss': tensor(0.3616, device='cuda:0'), 'epoch': 0}, {'loss': 0.48756, 'input': ..., 'ouput': ...}]`

Any ideas to solve this problem? I will appreciate your help! Here is the whole error stack:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-a4569c0ab5cd> in <module>
      3 logging.getLogger('allennlp').setLevel(logging.INFO)
      4 
----> 5 train("/shared/labs/workflow/configs/ner_span_path_2.jsonnet", "/Dropbox/EvidLabs/atee/lab_data_enrichment_project/train/DataEnrichmentSpanPath/DataEnrichmentSpanPath__INV_M_span_path_42_43", AllenNlpLightningModule, tags=[])
      6 
      7 # train("/shared/labs/workflow/configs/candidate_span_grouper.jsonnet", "./DataEnrichmentBunchedCrfTaggerTest", AllenNlpLightningModule, tags=["Frozen"])

/code/evid-research/evid2/pytorch_lightning/trainer/config_file_trainer.py in train(config_file, project_dir, lightning_module, run_id, run_name_override, tags, dry_run)
    171     # 3 START TRAINING
    172     # ------------------------
--> 173     trainer.fit(model)
    174 
    175 

/code/pytorch-lightning/pytorch_lightning/trainer/trainer.py in fit(self, model, train_dataloader, val_dataloaders)
    971         # easier to avoid NCCL issues
    972         elif self.use_dp:
--> 973             self.dp_train(model)
    974 
    975         elif self.use_horovod:

/code/pytorch-lightning/pytorch_lightning/trainer/distrib_parts.py in dp_train(self, model)
    265         model = LightningDataParallel(model, device_ids=device_ids)
    266 
--> 267         self.run_pretrain_routine(model)
    268 
    269         model.forward = model_autocast_original_forward

/code/pytorch-lightning/pytorch_lightning/trainer/trainer.py in run_pretrain_routine(self, model)
   1154 
   1155         # CORE TRAINING LOOP
-> 1156         self.train()
   1157 
   1158     def test(

/code/pytorch-lightning/pytorch_lightning/trainer/training_loop.py in train(self)
    368                 # RUN TNG EPOCH
    369                 # -----------------
--> 370                 self.run_training_epoch()
    371 
    372                 if self.max_steps and self.max_steps <= self.global_step:

/code/pytorch-lightning/pytorch_lightning/trainer/training_loop.py in run_training_epoch(self)
    480             # SAVE METRICS TO LOGGERS
    481             # -----------------------------------------
--> 482             self.save_train_loop_metrics_to_loggers(batch_idx, batch_output)
    483 
    484             # progress global step according to grads progress

/code/pytorch-lightning/pytorch_lightning/trainer/training_loop.py in save_train_loop_metrics_to_loggers(self, batch_idx, batch_output)
    564             # logs user requested information to logger
    565             print("batch_output.batch_log_metrics", batch_output.batch_log_metrics.keys())
--> 566             self.log_metrics(batch_output.batch_log_metrics, batch_output.grad_norm_dic)
    567 
    568     def save_loggers_in_training_loop(self, batch_idx):

/code/pytorch-lightning/pytorch_lightning/trainer/logging.py in log_metrics(self, metrics, grad_norm_dic, step)
     72         if self.is_global_zero and self.logger is not None:
     73             print("scalar_metrics", scalar_metrics.keys())
---> 74             self.logger.agg_and_log_metrics(scalar_metrics, step=step)
     75             self.logger.save()
     76 

/code/pytorch-lightning/pytorch_lightning/loggers/base.py in agg_and_log_metrics(self, metrics, step)
    131         """
    132         print("metrics", metrics.keys())
--> 133         agg_step, metrics_to_log = self._aggregate_metrics(metrics=metrics, step=step)
    134 
    135         if metrics_to_log:

/code/pytorch-lightning/pytorch_lightning/loggers/base.py in _aggregate_metrics(self, metrics, step)
     88 
     89         # compute the metrics
---> 90         agg_step, agg_mets = self._reduce_agg_metrics()
     91 
     92         # as new step received reset accumulator

/code/pytorch-lightning/pytorch_lightning/loggers/base.py in _reduce_agg_metrics(self)
    109             print("self._agg_key_funcs", self._agg_key_funcs)
    110             print("self._agg_default_func", self._agg_default_func)
--> 111             agg_mets = merge_dicts(self._metrics_to_agg, self._agg_key_funcs, self._agg_default_func)
    112         return self._prev_step, agg_mets
    113 

/code/pytorch-lightning/pytorch_lightning/loggers/base.py in merge_dicts(dicts, agg_key_funcs, default_func)
    386             print("fn or default_func", fn or default_func)
    387             print("values_to_agg", values_to_agg)
--> 388             d_out[k] = (fn or default_func)(values_to_agg)
    389 
    390     return d_out

<__array_function__ internals> in mean(*args, **kwargs)

/venvs/dev3.7/lib/python3.7/site-packages/numpy/core/fromnumeric.py in mean(a, axis, dtype, out, keepdims)
   3333 
   3334     return _methods._mean(a, axis=axis, dtype=dtype,
-> 3335                           out=out, **kwargs)
   3336 
   3337 

/venvs/dev3.7/lib/python3.7/site-packages/numpy/core/_methods.py in _mean(a, axis, dtype, out, keepdims)
    149             is_float16_result = True
    150 
--> 151     ret = umr_sum(arr, axis, dtype, out, keepdims)
    152     if isinstance(ret, mu.ndarray):
    153         ret = um.true_divide(

TypeError: unsupported operand type(s) for +: 'dict' and 'dict'
```

### To Reproduce

1. Add a training_epoch_end function to your Lightning Module and run it. You can use mine in the code sample section if you want. The key is that the "log"  section of training_epoch_end dictionary output must be of a different format than the dictionary that contains the results of your training sample while training (like in the example I provided `{'train_loss': tensor(0.3616, device='cuda:0'), 'epoch': 0}` and `{'loss': 0.48756, 'input': ..., 'ouput': ...}` are different formats as they don't share the same keys).

#### Code sample
```
 def training_epoch_end(self, outputs):
        print("training_epoch_end")
        print(len(outputs))

        prefix = "train_"
        metric_modules = self.training_metric_modules

        # Handle loss:
        avg_loss = torch.stack([x['loss'] for x in outputs]).mean()
        result = {'log': {}, 'progress_bar': {}}
        result[prefix + 'loss'] = avg_loss
        result['log'][prefix + 'loss'] = avg_loss
        result['progress_bar'][prefix + 'loss'] = avg_loss

        # Add tracking variables
        result['log']['epoch'] = self.current_epoch

        return result
```

### Expected behavior

I thought I would be able to run training_epoch_end function with no combination with the training samples in the next epoch. I expected no error, like running validation_epoch_end.

### Environment
* CUDA:
        - GPU:
                - Tesla V100-SXM2-16GB
        - available:         True
        - version:           10.2
* Packages:
        - numpy:             1.18.5
        - pyTorch_debug:     False
        - pyTorch_version:   1.5.0
        - pytorch-lightning: 0.8.4
        - tensorboard:       2.2.2
        - tqdm:              4.46.1
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                -
        - processor:         x86_64
        - python:            3.7.7
        - version:           #113-Ubuntu SMP Wed Jan 29 14:54:54 UTC 2020

### PR DESCRIPTION
**Pull request by ameliatqy**
## What does this PR do?

Fixes #2455. This used to be https://github.com/PyTorchLightning/pytorch-lightning/pull/2475 but I messed up on rebasing and to be safe, created a new pull request. This is a Draft PR.

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Yes, I am new to contributing to open-source and actually find it quite fun. May be time to consider a new hobby...

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -504,6 +505,9 @@ def run_training_epoch(self):
         # epoch end hook
         self.run_on_epoch_end_hook(model)
 
+        # increate global step by one to progress to the next epoch
+        self.global_step += 1
```
Comment (awaelchli) [inline pytorch_lightning/trainer/training_loop.py:509]: I think you also need to do do self.increment_accumulated_grad_global_step() here, otherwise the total_train_batch_idx does not get updated. I think that's why the tests are failing.

File: pytorch_lightning/loggers/base.py [hunk 2]
```diff
@@ -102,7 +102,24 @@ def _reduce_agg_metrics(self):
         elif len(self._metrics_to_agg) == 1:
             agg_mets = self._metrics_to_agg[0]
         else:
-            agg_mets = merge_dicts(self._metrics_to_agg, self._agg_key_funcs, self._agg_default_func)
+            # check if dictionary keys are unique
+            agg_keys = set([key for met in self._metrics_to_agg for key in met.keys()])
+            num_keys = sum([len(met) for met in self._metrics_to_agg])
+
+            # exclude 'epoch' because it is a metric automatically added in by log_metrics and will count as a
+            # duplicate. If you want to get rid of this, I would suggest you should get rid of `scalar_metrics[
```
Comment (awaelchli) [inline pytorch_lightning/loggers/base.py:110]: are the changes here really related to the global step issue?
If possible could we make a separate PR for this? we only want to fix one thing at a time.

Comment (ameliatqy) [reply pytorch_lightning/loggers/base.py:110]: I can see it being a separate issue. But, this piece of code is necessary for the code to run smoothly. Otherwise, when we fix the step issue (i.e. combining the last training batch + training_epoch_end + validation_epoch_end metrics in the same step), there will be an error combining the three pieces of data together because the original `merge_dict()` in `_reduce_agg_metrics()` requires that all the metric dictionaries have the same keys - which for these three types of data is usually not the case.

I don't mind doing a separate pull request for this though - but it will probably result in most tests failing for this pull request, which we will probably have to make sure to fix in the new pull request.

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (ameliatqy) [reply pytorch_lightning/trainer/training_loop.py]: I made the change and it works as expected. So I have changed it. 

However, the tests are still failing. Actually, I think I am confused about the test. I ran the local test on my pull request and got this total summary result at the end:
Screen Shot 2020-07-03 at 12 40 46 AM

I also ran the test over the original Pytorch Lightning master branch and also got this total summary result at the end:
Screen Shot 2020-07-03 at 12 34 03 AM

Am I making a mistake somewhere? Should I have based my branch off on something other than the master branch?

**Comment by Borda** [comment]
@ameliatqy I have rebased it to master, can we finish it in this release?

**Comment by ameliatqy** [comment]
> @ameliatqy I have rebased it to master, can we finish it in this release?

I have worked on the pull request and the checks have all passed. How do I get it approved? I  see that I need three approving reviews - am I responsible for finding people to review it or will the reviewers be assigned by someone else?

**Comment by awaelchli** [comment]
> > @ameliatqy I have rebased it to master, can we finish it in this release?
> 
> I have worked on the pull request and the checks have all passed. How do I get it approved? I see that I need three approving reviews - am I responsible for finding people to review it or will the reviewers be assigned by someone else?

Sorry for this late response.
Anyone in the community can review, but only the reviews of the contributors count towards a merge decision. 

I have two questions:
1. Does the issue with global step still exist on master?
2. Why do we need a decrement in the global_step? I would prefer a solution where this is not needed. Also, there was recently a fix #2853 in accumulate batches for the last batch, so I have the feeling this relates here.

**Comment by ameliatqy** [comment]
> > > @ameliatqy I have rebased it to master, can we finish it in this release?
> > 
> > 
> > I have worked on the pull request and the checks have all passed. How do I get it approved? I see that I need three approving reviews - am I responsible for finding people to review it or will the reviewers be assigned by someone else?
> 
> Sorry for this late response.
> Anyone in the community can review, but only the reviews of the core team count towards a merge decision.
> 
> I have two questions:
> 
> 1. Does the issue with global step still exist on master?
> 2. Why do we need a decrement in the global_step? I would prefer a solution where this is not needed. Also, there was recently a fix #2853 in accumulate batches for the last batch, so I have the feeling this relates here.

1. Yes, the issue still exists on master - I ran a test today and I still get the same error with the current master branch
2. I decremented it because, as discussed in issue #2455 , I plan to put the results of training_epoch_end in the same step as the results of the last batch when logging metrics. As such, the steps increment in the batch loop as usual until we get to the last step, in which we will increment the global_step at the very end of the function. The issue here is that we want to break the loop when `batch_idx >= self.num_training_batches`, even on the first loop. On the first loop, you can break the loop with no repercussions. But there are multiple loops, we would've incremented the global_step/total_batch_idx by one extra step - so the results of the last batch and training_epoch_end aren't stored together anymore. So I had to decrement it to the global_step/total_batch_idx of the last loop so that when` self.run_training_epoch_end()` is executed, the results of the last batch and training_epoch_end are stored together and the number of runs will sync together (as you described in https://github.com/PyTorchLightning/pytorch-lightning/issues/2455#issuecomment-653004576)

However, I understand you and I got rid of the decrement function - you can look at my new solution and see what you think. Additionally, the pull request that you linked for fixing accumulate batches is a different fix. That fixes a mechanism in which an action for the last batch wasn't executed when it was supposed to. This isn't the case here, here we can fixing a case in which a dictionary may have different keys (i.e. keys from the last batch training and keys from training_epoch_end). You can review my reasoning here which provides more details: https://github.com/PyTorchLightning/pytorch-lightning/issues/2455#issuecomment-653256670

**Comment by ameliatqy** [comment]
Just so you know, I had to remove ` self.logger.save()` from the function `log_metrics` because it was resetting `self._metrics_to_agg` every time log_metrics was called. This was not letting metrics from the same self.global_step aggregate and consequently making my epochs/step 'out of step' as talked about in https://github.com/PyTorchLightning/pytorch-lightning/issues/2455#issuecomment-653004576

After fixing that issues, I realize now that some tests aren't passing. Here is why:
```
FAILED tests/loggers/test_all.py::test_loggers_fit_test[TensorBoardLogger] - ...
FAILED tests/loggers/test_all.py::test_loggers_fit_test[CometLogger] - Assert...
FAILED tests/loggers/test_all.py::test_loggers_fit_test[MLFlowLogger] - Asser...
FAILED tests/loggers/test_all.py::test_loggers_fit_test[NeptuneLogger] - Asse...
FAILED tests/loggers/test_all.py::test_loggers_fit_test[TestTubeLogger] - Ass...
FAILED tests/loggers/test_all.py::test_loggers_fit_test[WandbLogger] - Assert...
```
These tests aren't passing because they expect the step of the test to be different from the validation and training. In order to keep the epoch and steps in lines (i.e. number of total steps = no batches*number of epochs), I believe the the test step should be the same as the training_epoch_end, validation_epoch_end and the results from the last batch.

```
FAILED tests/loggers/test_base.py::test_adding_step_key - assert 1 == 2
```
This test fails because it expects  validation_epoch_end and training_epoch_end to share the same step number. As discussed, this is not possible if we want to  keep the epoch and steps in lines (i.e. number of total steps = no batches*number of epochs).

So I guess I have a design question. Do you still want the epochs and step numbers to line up (i.e. number of total steps = no batches*number of epochs like in https://github.com/PyTorchLightning/pytorch-lightning/issues/2455#issuecomment-653004576)? Or should the training_epoch_end/validation_epoch_end/test_epoch_end have their own step. Please note that this will result in an offset as described in the comment linked (which I believe is undesired but let me know).

**Comment by williamFalcon** [comment]
please rebase master since this code is super old haha

**Comment by ameliatqy** [comment]
> please rebase master since this code is super old haha

Interesting, I thought I rebased my branch today XD I just rebased it again - does it look right? Rebasing is obviously not my strong point XD

**Comment by Borda** [comment]
@ameliatqy how about this one, mind rebase and resolve conflicts? 🐰

**Comment by ameliatqy** [comment]
> @ameliatqy how about this one, mind rebase and resolve conflicts? 🐰

Sure, I'd be happy to. But to fully resolve this pull request, could someone answer my question in https://github.com/PyTorchLightning/pytorch-lightning/pull/2478#issuecomment-675183459 ? I can fully finish up the pull request without an answer to this question.

File: pytorch_lightning/trainer/logging.py [hunk 3]
```diff
@@ -73,8 +74,6 @@ def log_metrics(self, metrics, grad_norm_dic, step=None):
         # log actual metrics
         if self.is_global_zero and self.logger is not None:
             self.logger.agg_and_log_metrics(scalar_metrics, step=step)
-            self.logger.save()
```
Comment (EspenHa) [inline pytorch_lightning/trainer/logging.py:76]: How is this related to https://github.com/PyTorchLightning/pytorch-lightning/issues/3398#issuecomment-691011447?
@SkafteNicki Maybe you know something about this also?

**Comment by awaelchli** [comment]
This is now fixed on master. There is no offset anymore and training epoch end logs at the correct step.
